tripal/includes/TripalFieldQuery.inc

```diff
@@ -32,49 +90,109 @@ class TripalFieldQuery extends EntityFieldQuery {
-          $results = $st_results;
-        }
-        // If other results already exist then we want to find the intersection
-        // of the two and only save those.
-        else {
-          $intersection = array(
-            'TripalEntity' => array(),
-          );
-          foreach ($st_results['TripalEntity'] as $entity_id => $stub) {
-            if (array_key_exists($entity_id, $results['TripalEntity'])) {
-              $intersection['TripalEntity'][$entity_id] = $stub;
-            }
-          }
-          $results = $intersection;
-        }
+        $results = $this->_intersectResults($results, $st_results);
       }
     }
-    // If there are no fields then default to the original
-    // EntityFieldQuery() functionality.
-    else {
-      $results = call_user_func($this->queryCallback(), $this);
+            
+    // If we have relationships, then handle those.
+    if (!empty($this->relationshipConditions)) {
+      $st_results = tripal_field_storage_query($this);
+      $results = $this->_intersectResults($results, $st_results);
+      
+      // Are there any other callbacks that need running other than
+      // propertyQuery as we would have run those with the relationships.
+      $callback = $this->queryCallback();
+      if ($callback and $callback[1] != 'propertyQuery' and 
+          $callback[1] != 'tripal_field_storage_query') {
+        $st_results = call_user_func($callback, $this);
+        $results = $this->_intersectResults($results, $st_results);
+      }
+    }
+    // There are no fields or relationships so just use the default 
+    // callback for the query.
+    else if (!$this->fields){
+      $callback = $this->queryCallback();
+      $st_results = call_user_func($callback, $this);
+      $results = $this->_intersectResults($results, $st_results);
     }
 
-    if ($results and $this->count) {
-      if (!is_numeric($results)) {
-        throw new Exception('Query callback function did not provide a numeric value: ' . $this->queryCallback());
+    // If this is a count query then it should return a numeric value.
+    if ($this->count) {
+      if (!$results or !is_array($results)) {
+        return 0;
       }
-      return $results;
+      return count($results['TripalEntity']);
+    }
+    return $results;
+  }
+  
+  /**
+   * Generates an intersection of results from different storage back-ends.
+   */
+  protected function _intersectResults($current_results, $new_results) {
+
+    // If we currently don't have any results then just allow all through.
+    // This is the first result set.
+    if (array_key_exists('first_results', $current_results)) {
+      return $new_results;
+    }
+    
+    // Iterate through all of the new results and only include those that
+    // exist in both the current and new.
+    $intersection = [];
+    foreach ($new_results['TripalEntity'] as $entity_id => $stub) {
+      if (array_key_exists($entity_id, $current_results['TripalEntity'])) {
```

Comment:
if i do a search that returns no results (IE pick a pub year that is wrong) I get the following errors at this line:


```
Notice: Undefined index: TripalEntity in TripalFieldQuery->_intersectResults() (line 148 of /Users/chet/UTK/tripal/sites/all/modules/tripal/tripal/includes/TripalFieldQuery.inc).
Warning: array_key_exists() expects parameter 2 to be array, null given in TripalFieldQuery->_intersectResults() (line 148 of /Users/chet/UTK/tripal/sites/all/modules/tripal/tripal/includes/TripalFieldQuery.inc).
Notice: Undefined index: TripalEntity in TripalFieldQuery->_intersectResults() (line 148 of /Users/chet/UTK/tripal/sites/all/modules/tripal/tripal/includes/TripalFieldQuery.inc).
Warning: array_key_exists() expects parameter 2 to be array, null given in TripalFieldQuery->_intersectResults() (line 148 of /Users/chet/UTK/tripal/sites/all/modules/tripal/tripal/includes/TripalFieldQuery.inc).
```